vyper/semantics/analysis/imports.py

```diff
@@ -0,0 +1,328 @@
+
+        if alias is None:
+            alias = node.name
+
+        # don't handle things like `import x.y`
+        if "." in alias:
+            msg = "import requires an accompanying `as` statement"
+            suggested_alias = node.name[node.name.rfind(".") :]
+            hint = f"try `import {node.name} as {suggested_alias}`"
+            raise StructureException(msg, node, hint=hint)
+
+        self._add_import(node, 0, node.name, alias)
+
+    def _handle_ImportFrom(self, node: vy_ast.ImportFrom):
+        # from m.n[module] import x[name] as y[alias]
+        alias = node.alias or node.name
+
+        module = node.module or ""
+        if module:
+            module += "."
+
+        qualified_module_name = module + node.name
+        self._add_import(node, node.level, qualified_module_name, alias)
+
+    def _add_import(
+        self, node: vy_ast.VyperNode, level: int, qualified_module_name: str, alias: str
+    ) -> None:
+        compiler_input, ast = self._load_import(node, level, qualified_module_name, alias)
+        node._metadata["import_info"] = ImportInfo(
+            alias, qualified_module_name, compiler_input, ast
+        )
+
+    # load an InterfaceT or ModuleInfo from an import.
+    # raises FileNotFoundError
+    def _load_import(self, node: vy_ast.VyperNode, level: int, module_str: str, alias: str) -> Any:
+        # the directory this (currently being analyzed) module is in
+        ast = self.graph.current_module
+        self_search_path = Path(ast.resolved_path).parent
+
+        with self.input_bundle.poke_search_path(self_search_path):
+            return self._load_import_helper(node, level, module_str, alias)
+
+    def _load_import_helper(
+        self, node: vy_ast.VyperNode, level: int, module_str: str, alias: str
+    ) -> tuple[CompilerInput, Any]:
+        if _is_builtin(module_str):
+            return _load_builtin_import(level, module_str)
+
+        path = _import_to_path(level, module_str)
+
+        if path in self.graph.imported_modules:
+            previous_import_stmt = self.graph.imported_modules[path]
+            raise DuplicateImport(f"{alias} imported more than once!", previous_import_stmt, node)
+
+        self.graph.imported_modules[path] = node
+
+        err = None
+
+        try:
+            path_vy = path.with_suffix(".vy")
+            file = self.input_bundle.load_file(path_vy)
+            assert isinstance(file, FileInput)  # mypy hint
+
+            module_ast = self._ast_from_file(file)
+            self.resolve_imports(module_ast)
+
+            return file, module_ast
+
+        except FileNotFoundError as e:
+            # escape `e` from the block scope, it can make things
+            # easier to debug.
+            err = e
+
+        try:
+            file = self.input_bundle.load_file(path.with_suffix(".vyi"))
+            assert isinstance(file, FileInput)  # mypy hint
+            module_ast = self._ast_from_file(file)
+
+            # no recursion yet
```

Comment:
bit unclear whether recursion comes in other parts of the code or whether this feature will be enabled in the future

Reply:
159efa1ce9f0ab0f62e14ea3755e84bc40f976bf